students/jesse_miller/session04/trigrams.py

```diff
@@ -16,55 +13,62 @@ def parse_file():
+    startline = 'Produced by an anonymous Project Gutenberg'\
+                'volunteer and Jose Menendez'
+    endline = 'End of the Project Gutenberg EBook'
 
-    for i in sherlock_input:
-        i = i.translate(translator).split(' ')
-        temp_dict = dict(i[l:l+2] for l in range(0, len(i) - 2))
-        sherlock_dict = dict(list(sherlock_dict.items()) +
-                             list(temp_dict.items()))
-    create_tuples(sherlock_dict)
+    with open('sherlock.txt') as input_file:
+        input_file = [line.rstrip('\n') for line in input_file]
+        full_list = []
+        start = 0
+        for line in input_file:
+            if line[:len(startline)] == startline:
+                start = 1
+                continue
+            if not start:
+                input_file = [line.rstrip('\n') for line in input_file]
+            if line[:len(endline)] == endline:
+                break
+    return full_list
 
+def make_words(input_file):
+    translator = str.maketrans('', '', string.punctuation)
+    full_list = []
+    for line in input_file:
+        full_list.extend(line.translate(translator).split(' '))
+    return full_list
 
-def create_tuples(sherlock_dict):
-    """
-    This should hopefully create the tuples needed for generating the trigrams
-    """
-    sherlock_tupled = {}
 
-    for i in range(len(sherlock_dict) - 2):
-        pair = ' '.join(dict(itertools.islice(sherlock_dict.items(), 2)))
-        follower = dict(itertools.islice(sherlock_dict.items(),i + 2))
-        #follower = sherlock_dict[i + 2]
-        if pair not in sherlock_tupled:  # Check if the pair is already in the dictionary
-            sherlock_tupled[pair] = [follower]
+def build_trigram(full_list):
+    sherlock_dict = {}
+    for i in range(0, len(full_list) - 2):
+        trigram_key = (full_list[i], full_list[i+1])
+        trigram_value = full_list[i+2]
+        if trigram_key in sherlock_dict:
+            sherlock_dict[trigram_key].append(trigram_value)
         else:
-            sherlock_tupled[pair].append(follower)
-    trigram_gen(sherlock_tupled)
-
-
-def trigram_gen(sherlock_tupled):
-    """
-    This, should assemble the randomized texts
-    """
-    v = 20
+            sherlock_dict[trigram_key] = [trigram_value]
+    return sherlock_dict
 
-    print("I am Sherlock reconstructed:")
-    for i in range(len(sherlock_tupled)):
-        text = ','.join(dict(itertools.islice(sherlock_tupled.items(), 2)) \
-        for _ in range(v))
-        #text = random.sample(list(sherlock_tupled), 8)
-        print(text)
+def build_text(sherlock_dict):
+    pass
 
 def main():
     """
     Main run list
     """
-    parse_file()
+    try:
+        filename = sys.argv[1]
+    except IndexError:
+        print("You must pass in a filename")
+        sys.exit(1)
+
+    in_data = parse_file()
+    words = make_words(input_file)
+    word_pairs = build_trigram(full_list)
+    new_text = build_text(sherlock_dict)
 
+    print(new_text)
```

Comment:
I changed _main()_ to the code below, to match the other suggested changes. Note that it ends by displaying the keys (2-word tuples) and values (lists with 1+ elements) of the trigram dictionary.  The next step will be to add some code to _build_text()_. The logic for that is somewhat simple:

1. Pick a random key in the trigram dictionary.
2. Get the list of words associated to that key.
3. Pick a random word from the list of words.
4. Add the two words from the key and the word from the list to the new text.
5. Create a new 2-word tuple formed by the second word of the current key and the word for the list. This will be key for the next iteration.
6. Get the list of words associated to the new key.
7. Select a random word from that list and add it to the new text.
8. Repeat steps 5-7 until you create a key that doesn't exist in the trigrams dictionary or run into another stop condition you define.

```
def main():
    """
    Main run list
    """
    try:
        filename = sys.argv[1]
    except IndexError:
        print("You must pass in a filename")
        sys.exit(1)

    input_file = parse_file(filename)
    words = make_words(input_file)
    word_pairs = build_trigram(words)
    for key, value in word_pairs.items():
        print(f'Key: {key}, Value: {value}')
```